src/common/buf.c

```diff
@@ -108,28 +105,38 @@ buf_add(struct buf *s, const char *data)
 void
 buf_add_char(struct buf *s, char ch)
 {
-	assert(s->buf);
-
-	if (s->alloc <= s->len + 1) {
-		s->alloc = s->alloc * 3 / 2 + 16;
-		s->buf = xrealloc(s->buf, s->alloc);
-	}
+	buf_expand(s, s->len + 1);
 	s->buf[s->len++] = ch;
 	s->buf[s->len] = '\0';
 }
 
 void
 buf_clear(struct buf *s)
 {
-	assert(s->buf);
-
-	s->len = 0;
-	s->buf[0] = '\0';
+	if (s->alloc) {
+		assert(s->buf);
+		s->len = 0;
+		s->buf[0] = '\0';
+	} else {
+		*s = BUF_INIT;
+	}
 }
 
 void
 buf_reset(struct buf *s)
 {
-	zfree(s->buf);
-	buf_init(s);
+	if (s->alloc) {
+		free(s->buf);
+	}
+	*s = BUF_INIT;
```

Comment:
Can we do this here? Would the `""` not be lost?

Reply:
String literals have static storage duration according to C99. Does that address the concern?

Reply:
I see what you mean. It just feels wrong.

Reply:
You know, I don't think I would have passed a technical interview question on string literals having static storage, but now that the penny has dropped I like it. Makes sense.
I'm not sure I've ever used this approach as a design pattern though - so that's quite exciting in itself.
Yes, agree it is particularly clean just doing `""`.
Let's go for it.